### ISSUE TITLE
improve resume from checkpoint

### ISSUE THREAD

**Issue by felipemello1** [pytorch/torchtune#1551]
The current experience with resume from checkpoint can improve. A few potential ways:

1. good defaults: Resuming from checkpoint should have as default using the last checkpoint saved, so the user can just turn resume_from_checkpoint=True, instead of having to pass paths;
2. always save when training ends: The user shouldnt have to train at least 2 epochs to resume from checkpoint. Training only 1 epoch should be fine. In the current state, 1 epoch or less is considered an intermediate ckpt and doesnt save recipe_state
3. resume from last epoch: currently we resume from epoch-1, which feels odd and (wastes 1 epoch of training?)
<img width="398" alt="image" src="https://github.com/user-attachments/assets/c2604710-4981-4a43-a34b-642c5f804bdb">
4.the user should have control over ckpt frequency, including no ckpt at all for when doing hyperparameter search or just testing ideas. Probably a good idea here what flexibility users have in other libraries

cc: @joecummings , tagging you because i have heard you were planning to work on this in october

**Comment by JinchuLi2002** [comment]
To add to this, currently resume from checkpoint does not work correctly with lr scheduler. For example, once I used 2e-4 lr with cosine scheduler, and the training was interrupted due to cluster connection issue a few epochs later (at a LR of 7e-5), there was no easy way to resume training with the new LR, even if I manually change the LR parameter from configs the training seemed to continue by using the original LR 2e-4.

**Comment by felipemello1** [comment]
thanks @JinchuLi2002 ! hopefully we can get to this issue soon.

**Comment by felipemello1** [comment]
just dumping info here

- default of torch.load will change to weights_only=True: https://dev-discuss.pytorch.org/t/bc-breaking-change-torch-load-is-being-flipped-to-use-weights-only-true-by-default-in-the-nightlies-after-137602/2573
- Theres a new library that makes loading weights 6x faster: https://x.com/_philschmid/status/1853339965612060797
- Our loading is naive and memory inefficient. Here is the right way of doing it: https://fb.workplace.com/groups/pytorch.dev/permalink/1643412869570520/
